# Autogen Tools

### Built-in Tools

https://microsoft.github.io/autogen/stable//user-guide/core-user-guide/components/tools.html

In [5]:
import os
import asyncio
from dotenv import load_dotenv
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.tools.http import HttpTool

In [9]:
# Load environment variables
load_dotenv()

api_key= os.getenv("OPENAI_API_KEY")
if not api_key:
    raise ValueError("Please set the OPENAI_API_KEY environment variable.")
print("Api Key has been setup")

Api Key has been setup


In [10]:
# Model Client

model_client= OpenAIChatCompletionClient(model="gpt-4o", api_key=api_key)

In [16]:
myagent= AssistantAgent(name="mychatbot", model_client=model_client)

In [ ]:
await myagent.run(task="what is agentiai")

TaskResult(messages=[TextMessage(id='aac5d440-7916-40d1-ac65-3939770ba432', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 8, 37, 19, 97573, tzinfo=datetime.timezone.utc), content='what is agentiai', type='TextMessage'), TextMessage(id='8262f23c-2412-4157-a6d4-f2212ee9eb6d', source='mychatbot', models_usage=RequestUsage(prompt_tokens=40, completion_tokens=240), metadata={}, created_at=datetime.datetime(2025, 7, 12, 8, 37, 27, 18280, tzinfo=datetime.timezone.utc), content='AgentiAI could refer to different things based on context, but generally, it appears to be a combination of "agent" and "AI," which might imply an artificial intelligence designed to act as an agent. In technology, an "agent" often refers to a software program or AI designed to perform specific tasks on behalf of a user.\n\nHere are a couple of possibilities:\n\n1. **AI Agent**: It can refer to a software entity that uses artificial intelligence to perform tasks autonomously, 

In [22]:
response= await myagent.run(task="what is agentiai")
print(response.messages[-1].content)

As of my last update in October 2023, "AgentiAI" is not recognized as a specific, widely-known entity, brand, or technology. It's possible that it could refer to a niche or newly emerging product, service, or company that has gained recognition after that time. Another possibility is that it could be a general reference to AI agents—software programs designed to perform tasks on behalf of users, often using artificial intelligence.

To find accurate and current information about "AgentiAI," I would recommend:

1. **Online Search**: Conduct a search using a search engine to see if there are any recent announcements or news articles about it.
   
2. **Social Media Exploration**: Platforms like LinkedIn, Twitter, or tech forums may have discussions or press releases related to new tech developments.

3. **Tech News Websites**: Websites dedicated to technology news often cover new products, companies, and innovations.

If you have more context about AgentiAI, such as where you encountered 

In [23]:
'''
{
  "fact": "Cats with long, lean bodies are more likely to be outgoing, and more protective and vocal than those with a stocky build.",
  "length": 121
}
'''

'\n{\n  "fact": "Cats with long, lean bodies are more likely to be outgoing, and more protective and vocal than those with a stocky build.",\n  "length": 121\n}\n'

In [25]:
schema= {
        "type": "object",
        "properties": {
            "fact": {
                "type": "string",
                "description": "A random cat fact"
            },
            "length": {
                "type": "integer",
                "description": "Length of the cat fact"
            }
        },
        "required": ["fact", "length"],
    }

schema

{'type': 'object',
 'properties': {'fact': {'type': 'string', 'description': 'A random cat fact'},
  'length': {'type': 'integer', 'description': 'Length of the cat fact'}},
 'required': ['fact', 'length']}

In [26]:
http_tool= HttpTool(
    name="cat_facts_api",
    description="get a cool cat fact",
    scheme="https",
    host="catfact.ninja",
    port=443,
    path="/fact",
    method="GET",
    return_type="json",
    json_schema= schema
)

In [28]:
agent= AssistantAgent(
    name="CatFactsAgent",
    model_client=model_client,
    system_message='You are a helpful assistant that can provide cat facts using the cat_facts_api tool. Give the result with summary',
    tools= [http_tool],
    reflect_on_tool_use=True
)

agent

In [29]:
result= await agent.run(task="Give me a random cat fact")

In [30]:
result.messages

[TextMessage(id='df2cbfa3-9d08-498d-b63e-58e411d42b46', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 10, 829732, tzinfo=datetime.timezone.utc), content='Give me a random cat fact', type='TextMessage'),
 ToolCallRequestEvent(id='129e2147-d9ac-48d9-a3d5-90b0bf14b038', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=98, completion_tokens=35), metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 11, 842975, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_zL0cFmUhCh2nlMrTEGIQdHix', arguments='{"fact":"Cats have five toes on their front paws but only four toes on their back paws.","length":71}', name='cat_facts_api')], type='ToolCallRequestEvent'),
 ToolCallExecutionEvent(id='7c290273-fc1c-4813-8892-cb0a63a9c49d', source='CatFactsAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 13, 805997, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content="{'


[TextMessage(id='df2cbfa3-9d08-498d-b63e-58e411d42b46', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 10, 829732, tzinfo=datetime.timezone.utc), content='Give me a random cat fact', type='TextMessage'),

 ToolCallRequestEvent(id='129e2147-d9ac-48d9-a3d5-90b0bf14b038', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=98, completion_tokens=35), metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 11, 842975, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_zL0cFmUhCh2nlMrTEGIQdHix', arguments='{"fact":"Cats have five toes on their front paws but only four toes on their back paws.","length":71}', name='cat_facts_api')], type='ToolCallRequestEvent'),

 ToolCallExecutionEvent(id='7c290273-fc1c-4813-8892-cb0a63a9c49d', source='CatFactsAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 13, 805997, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content="{'fact': 'Mohammed loved cats and reportedly his favorite cat, Muezza, was a tabby. Legend says that tabby cats have an “M” for Mohammed on top of their heads because Mohammad would often rest his hand on the cat’s head.', 'length': 210}", name='cat_facts_api', call_id='call_zL0cFmUhCh2nlMrTEGIQdHix', is_error=False)], type='ToolCallExecutionEvent'),

 TextMessage(id='653f4f2d-45ff-424b-898e-4dcaaa0c59c5', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=146, completion_tokens=59), metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 49, 15, 338129, tzinfo=datetime.timezone.utc), content="Here's an interesting cat fact: Mohammed reportedly loved cats, and his favorite cat was a tabby named Muezza. Legend has it that tabby cats have an “M” on their foreheads as a tribute to Mohammed, who was said to rest his hand on the cat’s head.", type='TextMessage')]

In [31]:
agent= AssistantAgent(
    name="CatFactsAgent",
    model_client=model_client,
    system_message='You are a helpful assistant that can provide cat facts using the cat_facts_api tool. Give the result with summary',
    tools= [http_tool],
    reflect_on_tool_use=False
)

agent

In [32]:
result= await agent.run(task="Give me a random cat fact")

In [33]:
result.messages

[TextMessage(id='147f057a-e54f-43d7-84f6-e48b3ed126b1', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 17, 770441, tzinfo=datetime.timezone.utc), content='Give me a random cat fact', type='TextMessage'),
 ToolCallRequestEvent(id='9c922174-4e15-410f-9fc9-15ea86c30acc', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=98, completion_tokens=36), metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 18, 830718, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_d2laIcmaKkY1gaklbAYvYkju', arguments='{"fact":"Cats have five toes on their front paws, but only four toes on their back paws.","length":68}', name='cat_facts_api')], type='ToolCallRequestEvent'),
 ToolCallExecutionEvent(id='2b2362a6-b99a-4535-8f27-c2a62cc9d991', source='CatFactsAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 19, 554802, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content="{

[TextMessage(id='147f057a-e54f-43d7-84f6-e48b3ed126b1', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 17, 770441, tzinfo=datetime.timezone.utc), content='Give me a random cat fact', type='TextMessage'),

 ToolCallRequestEvent(id='9c922174-4e15-410f-9fc9-15ea86c30acc', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=98, completion_tokens=36), metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 18, 830718, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_d2laIcmaKkY1gaklbAYvYkju', arguments='{"fact":"Cats have five toes on their front paws, but only four toes on their back paws.","length":68}', name='cat_facts_api')], type='ToolCallRequestEvent'),

 ToolCallExecutionEvent(id='2b2362a6-b99a-4535-8f27-c2a62cc9d991', source='CatFactsAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 19, 554802, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content="{'fact': 'Tylenol and chocolate are both poisionous to cats.', 'length': 50}", name='cat_facts_api', call_id='call_d2laIcmaKkY1gaklbAYvYkju', is_error=False)], type='ToolCallExecutionEvent'),
 
 ToolCallSummaryMessage(id='d7fa9e29-2d6f-4bc9-a920-824d49f7f941', source='CatFactsAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 9, 51, 19, 555800, tzinfo=datetime.timezone.utc), content="{'fact': 'Tylenol and chocolate are both poisionous to cats.', 'length': 50}", type='ToolCallSummaryMessage', tool_calls=[FunctionCall(id='call_d2laIcmaKkY1gaklbAYvYkju', arguments='{"fact":"Cats have five toes on their front paws, but only four toes on their back paws.","length":68}', name='cat_facts_api')], results=[FunctionExecutionResult(content="{'fact': 'Tylenol and chocolate are both poisionous to cats.', 'length': 50}", name='cat_facts_api', call_id='call_d2laIcmaKkY1gaklbAYvYkju', is_error=False)])]

In [34]:
agent= AssistantAgent(
    name="CatFactsAgent",
    model_client=model_client,
    system_message='You are a helpful assistant that can provide cat facts using the cat_facts_api tool. Give the result with summary',
    tools= [http_tool],
    reflect_on_tool_use=True
)

agent

In [35]:
result= await agent.run(task="write a python program for fibonacci series")

In [36]:
result.messages

[TextMessage(id='f5431813-711d-4a4f-8e9a-6e285149c0d2', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 12, 10, 5, 49, 978082, tzinfo=datetime.timezone.utc), content='write a python program for fibonacci series', type='TextMessage'),
 TextMessage(id='38d3c72b-8608-4c8b-8fe5-31731621ff43', source='CatFactsAgent', models_usage=RequestUsage(prompt_tokens=99, completion_tokens=186), metadata={}, created_at=datetime.datetime(2025, 7, 12, 10, 5, 52, 487019, tzinfo=datetime.timezone.utc), content='Below is a simple Python program to generate the Fibonacci series up to `n` terms:\n\n```python\ndef fibonacci_series(n):\n    fib_sequence = []\n    a, b = 0, 1\n    for _ in range(n):\n        fib_sequence.append(a)\n        a, b = b, a + b\n    return fib_sequence\n\n# Get the number of terms from the user\nn_terms = int(input("Enter the number of terms: "))\n\n# Generate and print the Fibonacci series\nresult = fibonacci_series(n_terms)\nprint(f"Fibonacci ser

In [38]:
print(result.messages[-1].content)

Below is a simple Python program to generate the Fibonacci series up to `n` terms:

```python
def fibonacci_series(n):
    fib_sequence = []
    a, b = 0, 1
    for _ in range(n):
        fib_sequence.append(a)
        a, b = b, a + b
    return fib_sequence

# Get the number of terms from the user
n_terms = int(input("Enter the number of terms: "))

# Generate and print the Fibonacci series
result = fibonacci_series(n_terms)
print(f"Fibonacci series up to {n_terms} terms: {result}")
```

This program defines a function `fibonacci_series(n)` that generates a list containing the first `n` terms of the Fibonacci sequence. It uses a simple iterative approach with two variables `a` and `b` to calculate the sequence, starting from 0 and 1.
